core/src/main/java/org/apache/calcite/sql/SqlDialect.java

```diff
@@ -1218,6 +1240,50 @@ public static String replace(
     }
   }
 
+  protected String getDateTimeFormatString(
+      String standardDateFormat, Map<SqlDateTimeFormat, String> dateTimeFormatMap) {
+    Pair<List<String>, List<Character>> dateTimeTokensWithSeparators =
+        getDateTimeTokensWithSeparators(standardDateFormat);
+    return getFinalFormat(dateTimeTokensWithSeparators.left,
+        dateTimeTokensWithSeparators.right, dateTimeFormatMap);
+  }
+
+  private Pair<List<String>, List<Character>> getDateTimeTokensWithSeparators(
+      String standardDateFormat) {
+    List<String> dateTimeTokens = new ArrayList<>();
+    List<Character> separators = new ArrayList<>();
+    int startIndex = 0;
+    int lastIndex = standardDateFormat.length() - 1;
+    for (int i = 0; i <= lastIndex; i++) {
+      if (i == lastIndex) {
+        dateTimeTokens.add(standardDateFormat.substring(startIndex));
```

Comment:
can we modify the logic? we can apply the loop from 0 to less than lastIndex and separate the if logic from the loop

Reply:
done